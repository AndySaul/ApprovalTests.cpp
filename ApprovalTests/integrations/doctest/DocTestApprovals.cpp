#include "ApprovalTests/integrations/doctest/DocTestApprovals.h"

namespace ApprovalTests
{

}
