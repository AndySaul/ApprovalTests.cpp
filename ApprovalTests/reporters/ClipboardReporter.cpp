#include "ApprovalTests/reporters/ClipboardReporter.h"

namespace ApprovalTests
{

}
