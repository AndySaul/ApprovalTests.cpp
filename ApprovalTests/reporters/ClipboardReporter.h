#ifndef APPROVALTESTS_CPP_COMMANDLINEREPORTER_H
#define APPROVALTESTS_CPP_COMMANDLINEREPORTER_H

#include "Reporter.h"
#include "../SystemUtils.h"


#include <string>
#include <iostream>

namespace ApprovalTests {
class ClipboardReporter : public Reporter {
public:
    static std::string getCommandLineFor(std::string received, std::string approved, bool isWindows)
    {
        if (isWindows) {
            return std::string("move /Y ") + "\"" + received + "\" \"" + approved + "\"";
        } else {
            return std::string("mv ") + "\"" + received + "\" \"" + approved + "\"";
        }
    }

    virtual bool report(std::string received, std::string approved) const override
    {
        copyToClipboard(getCommandLineFor(received, approved, SystemUtils::isWindowsOs()));
        return true;
    }

    static void copyToClipboard(const std::string& newClipboard) {
        /*
         This will probably not work on Linux.

         From https://stackoverflow.com/a/750466/104370
         In case of X, yes, there's xclip (and others). xclip -selection c will send data to the clipboard that
         works with Ctrl-C, Ctrl-V in most applications.

         If you're trying to talk to the Mac OS X clipboard, there's pbcopy.

         If you're in Linux terminal mode (no X) then maybe you need to look into gpm.

         There's also GNU screen which has a clipboard. To put stuff in there, look at the screen command "readreg".

         Under Windows/cygwin, use /dev/clipboard or clip for newer versions of Windows (at least Windows 10).
         */

        const std::string clipboardCommand = SystemUtils::isWindowsOs() ? "clip" : "pbclip";
        auto cmd = std::string("echo ") + newClipboard + " | " + clipboardCommand;
        system(cmd.c_str());
    }
};
}

#endif //APPROVALTESTS_CPP_COMMANDLINEREPORTER_H
