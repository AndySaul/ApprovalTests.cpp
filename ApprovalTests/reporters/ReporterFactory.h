#pragma once

#include "ApprovalTests/core/Reporter.h"

#include <memory>
#include <string>

namespace ApprovalTests
{
    class ReporterFactory
    {
    public:
        std::unique_ptr<Reporter> createReporter(const std::string& reporterName);
    };
}
