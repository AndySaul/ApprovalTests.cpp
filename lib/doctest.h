#include "doctest.2.4.beta.hpp"
