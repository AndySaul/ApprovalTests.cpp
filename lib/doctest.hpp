#include "doctest.2.3.3.hpp"
