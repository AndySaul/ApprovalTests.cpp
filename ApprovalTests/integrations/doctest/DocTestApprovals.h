#ifndef APPROVALTESTS_CPP_DOCTESTAPPROVALS_H
#define APPROVALTESTS_CPP_DOCTESTAPPROVALS_H

#include "../ApprovalTests/namers/ApprovalTestNamer.h"

// <SingleHpp unalterable>
#ifdef APPROVALS_DOCTEST

#define DOCTEST_CONFIG_IMPLEMENT
#include <doctest.hpp>

struct DocTestApprovalListener : doctest::ConsoleReporter
{
    TestName currentTest;

    // constructor has to accept the ContextOptions by ref as a single argument
    DocTestApprovalListener(const doctest::ContextOptions& in) : ConsoleReporter(in)
    {
    }

    void test_case_start(const doctest::TestCaseData& testInfo) override
    {
        ConsoleReporter::test_case_start(testInfo);

        currentTest.sections.push_back(testInfo.m_name);
        currentTest.setFileName(testInfo.m_file);
        ApprovalTestNamer::currentTest(&currentTest);
    }

    void test_case_end(const doctest::CurrentTestCaseStats& in) override
    {
        ConsoleReporter::test_case_end(in);

        while (!currentTest.sections.empty()) {
            currentTest.sections.pop_back();
        }
    }

    void subcase_start(const doctest::SubcaseSignature &signature) override
    {
        ConsoleReporter::subcase_start(signature);

        currentTest.sections.push_back(signature.m_name);
    }

    void subcase_end() override
    {
        ConsoleReporter::subcase_end();

        currentTest.sections.pop_back();
    }
};

REGISTER_REPORTER("approvals", 0, DocTestApprovalListener);

int main(int argc, char** argv)
{
    const char *myargv[] = {"program name", "--reporters=approvals", NULL};
    int myargc = sizeof(myargv) / sizeof(char*) - 1;

    return doctest::Context(myargc, myargv).run();
}

#endif // APPROVALS_DOCTEST
// </SingleHpp>
#endif //APPROVALTESTS_CPP_DOCTESTAPPROVALS_H
