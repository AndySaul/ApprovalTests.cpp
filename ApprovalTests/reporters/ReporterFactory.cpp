#include "ReporterFactory.h"

#include "ApprovalTests/reporters/AutoApproveIfMissingReporter.h"
#include "ApprovalTests/reporters/AutoApproveReporter.h"
#include "ApprovalTests/reporters/BlockingReporter.h"
#include "ApprovalTests/reporters/CIBuildOnlyReporter.h"
#include "ApprovalTests/reporters/ClipboardReporter.h"
#include "ApprovalTests/reporters/CombinationReporter.h"
#include "ApprovalTests/reporters/CommandReporter.h"
#include "ApprovalTests/reporters/CustomReporter.h"
#include "ApprovalTests/reporters/DefaultFrontLoadedReporter.h"
#include "ApprovalTests/reporters/DefaultReporter.h"
#include "ApprovalTests/reporters/DiffReporter.h"
#include "ApprovalTests/reporters/EnvironmentVariableReporter.h"
#include "ApprovalTests/reporters/FirstWorkingReporter.h"
#include "ApprovalTests/reporters/FrontLoadedReporterDisposer.h"
#include "ApprovalTests/reporters/GenericDiffReporter.h"
#include "ApprovalTests/reporters/LinuxReporters.h"
#include "ApprovalTests/reporters/MacReporters.h"
#include "ApprovalTests/reporters/QuietReporter.h"
#include "ApprovalTests/reporters/TextDiffReporter.h"
#include "ApprovalTests/reporters/WindowsReporters.h"

#include <map>
#include <functional>

#define APPROVAL_TESTS_REGISTER_REPORTER(name)                                           \
    map[#name] = []() { return std::unique_ptr<Reporter>(new name); }

namespace ApprovalTests
{
    std::unique_ptr<Reporter>
    ReporterFactory::createReporter(const std::string& reporterName)
    {
        std::map<std::string, std::function<std::unique_ptr<Reporter>()>> map;

        APPROVAL_TESTS_REGISTER_REPORTER(AutoApproveIfMissingReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(AutoApproveReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(CIBuildOnlyReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(ClipboardReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(DefaultFrontLoadedReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(DefaultReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(DiffReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(EnvironmentVariableReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(QuietReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(TextDiffReporter);

        APPROVAL_TESTS_REGISTER_REPORTER(Linux::BeyondCompareReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Linux::MeldReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Linux::SublimeMergeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Linux::KDiff3Reporter);

        APPROVAL_TESTS_REGISTER_REPORTER(Mac::AraxisMergeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Mac::BeyondCompareReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Mac::DiffMergeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Mac::KaleidoscopeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Mac::P4MergeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Mac::SublimeMergeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Mac::KDiff3Reporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Mac::TkDiffReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Mac::VisualStudioCodeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Mac::CLionDiffReporter);

        APPROVAL_TESTS_REGISTER_REPORTER(Windows::TortoiseDiffReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Windows::TortoiseGitDiffReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Windows::BeyondCompareReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Windows::WinMergeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Windows::AraxisMergeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Windows::CodeCompareReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Windows::SublimeMergeReporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Windows::KDiff3Reporter);
        APPROVAL_TESTS_REGISTER_REPORTER(Windows::VisualStudioCodeReporter);

        std::vector<std::string> candidateNames = {
            reporterName,
            // Allow program names to be specified without Reporter suffix
            reporterName + "Reporter"};

        for (auto& candidateName : candidateNames)
        {
            auto iter = map.find(candidateName);
            if (iter != map.end())
            {
                return iter->second();
            }
        }

        return std::unique_ptr<Reporter>();
    }
}
