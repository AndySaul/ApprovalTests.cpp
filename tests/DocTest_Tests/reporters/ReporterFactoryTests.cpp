#include "ApprovalTests/reporters/ReporterFactory.h"
#include "ApprovalTests/reporters/ClipboardReporter.h"

#include "doctest/doctest.h"

using namespace ApprovalTests;

TEST_CASE("ReporterFactory creates a Reporter given a string")
{
    ReporterFactory factory;
    std::unique_ptr<Reporter> reporter = factory.createReporter("ClipboardReporter");
    CHECK(reporter);
    CHECK(dynamic_cast<ClipboardReporter*>(reporter.get()));
}
