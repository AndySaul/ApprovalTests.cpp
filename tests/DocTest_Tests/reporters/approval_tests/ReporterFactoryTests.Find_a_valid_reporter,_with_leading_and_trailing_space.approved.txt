Kaleidoscope => Mac::KaleidoscopeReporter
 Kaleidoscope => Mac::KaleidoscopeReporter
Kaleidoscope  => Mac::KaleidoscopeReporter
 Kaleidoscope  => Mac::KaleidoscopeReporter

